torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -0,0 +1,281 @@
+  return !workException_;
+}
+
+bool ProcessGroupMPI::WorkMPI::wait() {
+  std::unique_lock<std::mutex> lock(workMutex_);
+  while (!completed_) {
+    workCV_.wait(lock);
+  }
+  return isSuccess();
+}
+
+void ProcessGroupMPI::WorkMPI::finish() {
+  {
+    std::unique_lock<std::mutex> lock(workMutex_);
+    completed_ = true;
+  }
+  workCV_.notify_all();
+}
+
+void ProcessGroupMPI::WorkMPI::finishWithException(
+    std::exception_ptr caughtWorkException) {
+  {
+    std::unique_lock<std::mutex> lock(workMutex_);
+    completed_ = true;
+    workException_ = caughtWorkException;
+  }
+  workCV_.notify_all();
+}
+
+const std::exception& ProcessGroupMPI::WorkMPI::exception() const {
+  try {
+    std::rethrow_exception(workException_);
+  } catch (const std::exception& e) {
+    return e;
+  }
+}
+
+// ProcessGroupMPI
+
+// Static global states
+int ProcessGroupMPI::numProcessGroups_ = 0;
+int ProcessGroupMPI::mpiThreadSupport_ = 0;
+bool ProcessGroupMPI::mpiInitialized_ = false;
+std::mutex ProcessGroupMPI::pgGlobalMutex_;
+
+ProcessGroupMPI::ProcessGroupMPI() : ProcessGroup(-1, 0), stop_(false) {
+  // Initialize MPI environment
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  // We only want to initialize once
+  if (!mpiInitialized_) {
+    MPI_CHECK(MPI_Init_thread(
+        nullptr, nullptr, MPI_THREAD_MULTIPLE, &mpiThreadSupport_));
+    if (mpiThreadSupport_ < MPI_THREAD_SERIALIZED) {
+      throw std::runtime_error(
+          "Used MPI implementation doesn't have the "
+          "minimum level of threading support: "
+          "MPI_THREAD_SERIALIZED. This is required by "
+          "c10d package");
+    }
+    if (std::atexit(mpiExit)) {
+      throw std::runtime_error("Fail to register the MPI exit handler");
+    }
+    mpiInitialized_ = true;
+  }
+
+  // Update the world size and rank
+  MPI_CHECK(MPI_Comm_size(MPI_COMM_WORLD, &size_));
+  MPI_CHECK(MPI_Comm_rank(MPI_COMM_WORLD, &rank_));
+
+  if (mpiThreadSupport_ != MPI_THREAD_MULTIPLE && numProcessGroups_ >= 1) {
+    throw std::runtime_error(
+        "More than one process group created, "
+        "this is not supported due to the used MPI "
+        "implementation doesn't provide the full support "
+        "of multi-threading");
+  }
+  // increase the total PG count
+  ++numProcessGroups_;
+
+  // Start the worker thread accepting MPI calls
+  workerThread_ = std::thread(&ProcessGroupMPI::runLoop, this);
+}
+
+ProcessGroupMPI::~ProcessGroupMPI() {
+  destroy();
+}
+
+void ProcessGroupMPI::destroy() {
+  std::unique_lock<std::mutex> lock(pgMutex_);
+
+  while (!queue_.empty()) {
+    queueConsumeCV_.wait(lock);
+  }
+  // Queue is empty, signal stop
+  stop_ = true;
+
+  // Release lock to allow threads to terminate
+  queueProduceCV_.notify_all();
+
+  lock.unlock();
+
+  // Join the single worker thread
+  workerThread_.join();
+
+  // Decrease the number of PG created
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  --numProcessGroups_;
```

Comment:
Would this allow create + destroy followed by create + destroy?

Reply:
I believe it will since I tested it, will double check